rfc/src/rfcs/0012-loop-contracts.md

```diff
@@ -0,0 +1,235 @@
+    #[kani::loop_modifies(x)]
+    while x > 1{
+        x = x - 1;
+    };
+
+    assert!(x == 1);
+}
+```
+write to only `x` in the loop, hence the `modifies` clause contains only `x`.
+Then when use the loop contracts to abstract the loop, Kani will only havoc the memory
+location `x` and keep `y` unchanged. 
+
+
+Kani will also verify if all writing targets in the loop are included in the `modifies`
+clause.
+
+
+Note that the `modifies` clause is optional, and Kani will infer the write set if not
+provided.
+
+## Detailed Design
+
+
+Kani implements the functionality of loop contracts in three places.
+
+1. Procedural macros `loop_invariant` and `loop_modifies`.
+2. Code generation for builtin functions expanded from the above two macros.
+3. GOTO-level loop contracts using CBMC's contract language generated in
+   `kani-compiler` for `loop-modifies` clauses.
+
+### Procedural macros `loop_invariant` and `loop_modifies`.
+
+The `loop_invariant` macro perform code generation for the loop invariant clause.
+The generated code consists of two parts: 
+1. a closure definition to wrap the loop invariant, which is an boolean expression.
+2. a call to a builtin function `kani_loop_invariant` at end of the loop.
+
+As an example, in the above program, the following code will be generated for the
+loop invariants clauses:
+
+```rs
+fn main() {
+    let mut x: u64 = kani::any_where(|i| *i >= 1);
+
+    let _loop_invariant_closure = || x >= 1;
+    while x > 1{
+        x = x - 1;
+        kani_loop_invariant_begin();
+        _loop_invariant_closure()
+        kani_loop_invariant_end();
+    };
+
+    assert!(x == 1);
+}
+```
+
+Similarly, we generate a call to the builtin function `kani_loop_modifies` for modifies
+clauses.
+
+
+### Code Generation for Builtin Functions
+
+When generating GOTO program from MIR, Kani will first scan for the placeholder function
+calls `kani_loop_invariant_begin` and `kani_loop_invariant_end` in the MIR. Then Kani
+will generate the corresponding GOTO-level statement expression for all instructions
+between the two placeholder function calls. At last, Kani will add the statement expression
+to the loop latch---the jump back to the loop head.
+
+The artifact `goto-instrument` in CBMC will extract the loop contracts from the named-subs
+of the loop latch, and the apply and prove the extracted loop contracts.
+
+Similarly, Kani will add the `modifies` targets into the named-subs of the loop latch for
+CBMC to extract and prove the loop contracts.
+
+
+### GOTO-Level Havocing
+
+The ordinary havocing in CBMC is not aware of the type constraints of Rust type.
+Hence, we will use customized havocing functions for modifies targets. In detail,
+Kani will generate code for the definition of corresponding `kani::any()` functions
+for each modifies target. Then Kani will create a map from the modifies target to the
+the name of its `kani::any()` function, and add the map to the loop latch too.
```

Comment:
Are you adding a new property to goto-program that will allow us to provide that information, or would this be a separate artifact that we will need to pass to `goto-instrument`? Can we expand that to handle havoc in contracts modifies?

Reply:
One thing that may be worth calling out is that it is perfectly fine for a loop to break the safety invariant of a value. So calling `kani::any()` could potentially be unsound if the loop is inside an unsafe block / function. If that is the case, you may need something in between. I.e., something that preserves validity invariant, but not safety invariant.

The UX question becomes how will the user specify the different scenarios. Do we create different modifies clauses? Or do we force the user to state the invariant that the modified type is safe?

Reply:
I think it should be consistent as how `kani::any()` works in an unsafe block. I propose to allow users provide custom `any()` functions for each modify target.

Reply:
Users don't really use `kani::any()` in code outside the harness / stubs. I suggest that you leave this as an open question for now.